### PR TITLE
Updated validation code to support validators changing aguments

### PR DESCRIPTION
**Pull request by Onager**
Apologies for the size of the change.

This is prepatory to changing the datetime validator to create datetime objects for modules to use

### CODE REVIEW

File: dftimewolf/cli/dftimewolf_recipes.py [hunk 1]
```diff
@@ -110,20 +111,23 @@ def __init__(
     self._data_files_path = ''
     self._recipes_manager = recipes_manager.RecipesManager()
     self._recipe = {}  # type: Dict[str, Any]
-    self._args_validator = args_validator.ValidatorManager()
+    self._state = None  # type: Optional[DFTimewolfState]
+    self._command_line_options = argparse.Namespace()
     self.dry_run = False
     self.cdm = cdm
-    self._state: "dftw_state.DFTimewolfState" # for pytype
+    # self._state: "dftw_state.DFTimewolfState" # for pytype
```
[the remark this answers is not in the document]
Comment (Onager) [reply dftimewolf/cli/dftimewolf_recipes.py:118]: Done

File: dftimewolf/lib/errors.py [hunk 2]
```diff
@@ -43,6 +43,44 @@ class CommandLineParseError(DFTimewolfError):
...
+  Attributes:
+    switch (str): The name of the argument that is invalid.
+    argument_value (str): The value of the argument that is invalid.
+    description (str): Description of why the argument is invalid.
+    message (str): The error message.
+    name (str): Name of the module that generated the error.
+    stacktrace (Optional[str]): Stacktrace leading to the error.
+    critical (Optional[bool]): Whether the error is critical or not. Critical
+        errors interrupt the recipe execution flow.
+    unexpected (Optional[bool]): Whether the error is unexpected.
```
[the remark this answers is not in the document]
Comment (Onager) [reply dftimewolf/lib/errors.py:58]: Done, updated the other docstring in this file too

File: tests/lib/args_validator.py [hunk 3]
```diff
@@ -217,28 +247,31 @@ class SubnetValidatorTest(unittest.TestCase):
...
-    params = {'comma_separated': True}
 
     for value in values:
-      val, msg = self.validator.Validate(value, params)
-      self.assertFalse(val)
-      self.assertEqual(msg, f'{value} is not a valid subnet.')
+      with self.assertRaisesRegex(
+          errors.RecipeArgsValidationFailure,
+          'Not a valid subnet'):
+        _ = self.validator.Validate(value, self.recipe_argument)
```
[the remark this answers is not in the document]
Comment (Onager) [reply tests/lib/args_validator.py:274]: Done

File: tests/lib/args_validator.py [hunk 4]
```diff
@@ -443,100 +485,102 @@ def test_ValidateSuccess(self):
...
-  def test_ValidationFailure(self):
-    """Tests validation failure."""
-    val, msg = self.vm.Validate('invalid',
-                                {'format': 'subnet', 'comma_separated': False})
-    self.assertFalse(val)
-    self.assertEqual(msg, 'invalid is not a valid subnet.')
 
-  def test_InvalidValidator(self):
-    """Tests an exception is thrown if an invalid validator is specified."""
+  def testRegisterModules(self):
```
[the remark this answers is not in the document]
Comment (Onager) [reply tests/lib/args_validator.py:550]: Done

File: dftimewolf/cli/dftimewolf_recipes.py [hunk 5]
```diff
@@ -310,28 +314,26 @@ def ValidateArguments(self) -> None:
...
+      argument_mandatory = switch == arg.switch
+      argument_set = switch in self.state.command_line_options
+
+      if argument_mandatory or argument_set:
+        argument_value = self.state.command_line_options[switch]
+        try:
+          valid_value = args_validator.ValidatorsManager.Validate(
+              str(argument_value), arg)
+          self.state.command_line_options[switch] = valid_value
+        except errors.RecipeArgsValidationFailure as exception:
```
[the remark this answers is not in the document]
Comment (Onager) [reply dftimewolf/cli/dftimewolf_recipes.py:329]: A failure means the argument is wrong/invalid, an error is something wrong with the validation code or the validation configuration.

I've cleaned up one usage that was not consistent with this definition.

**Comment by tomchop** [comment]
LGTM!

**Review by tomchop — APPROVED** [review]

**Pull request merged**